test/test_dynamic_transform.cpp

```diff
@@ -0,0 +1,485 @@
+
+    std::cout << "Before fusion concretization\n";
+    fusion.printMath();
+    std::cout << std::endl;
+
+    DynamicTransform::concretizeFusion(&fusion, info);
+
+    std::cout << "After fusion concretization\n";
+    fusion.printMath();
+    std::cout << std::endl;
+
+    TORCH_CHECK(
+        !fusion.hasDynamicTransform(), "Expected to have no dynamic transform");
+
+    fusion.printMath();
+  }
+}
+
+// Reshape of reshape
+TEST_F(NVFuserTest, DynamicTransform6_CUDA) {
+  std::vector<std::vector<std::vector<int64_t>>> reshape_lists = {
+      {{4, 3}, {3, 4}},
+      {{4, 3}, {3, 4}, {12}},
+      {{4, 3}, {3, 1, 4}, {12, 1}},
+      {{4, 3}, {12}, {3, 4}},
+      {{4, 3}, {1, 2, 1, 3, 2}, {3, 4}},
+  };
+
+  for (auto reshape_list : reshape_lists) {
+    std::cerr << "Reshape list: " << reshape_list << std::endl;
+    std::vector<TensorView*> reshape_tvs;
+
+    Fusion fusion;
+    FusionGuard fg(&fusion);
+
+    auto tv0 = makeSymbolicTensor(reshape_list.at(0).size());
+    fusion.addInput(tv0);
+
+    reshape_tvs.push_back(tv0);
+
+    for (auto it = reshape_list.begin() + 1; it != reshape_list.end(); ++it) {
+      auto shape = *it;
+      std::vector<Val*> shape_arg;
+      for (const auto i : c10::irange(shape.size())) {
+        (void)i;
+        shape_arg.push_back(IrBuilder::create<Int>());
+      }
+
+      auto tv = reshape(reshape_tvs.back(), shape_arg);
+      reshape_tvs.push_back(tv);
+    }
+    fusion.addOutput(reshape_tvs.back());
+
+    fusion.printMath();
+
+    ExpressionEvaluator expr_eval;
+
+    for (const auto i : c10::irange(reshape_list.size())) {
+      const auto& shape = reshape_list.at(i);
+      for (const auto j : c10::irange(shape.size())) {
+        expr_eval.bind(reshape_tvs.at(i)->axis(j)->extent(), shape.at(j));
+      }
+    }
+
+    auto info = DynamicTransform::getConcretizationInfo(&fusion, &expr_eval);
+
+    std::cerr << info.toString() << std::endl;
+
+    std::cout << "Before fusion concretization\n";
+    fusion.printMath();
+    std::cout << std::endl;
+
+    DynamicTransform::concretizeFusion(&fusion, info);
+
+    std::cout << "After fusion concretization\n";
+    fusion.printMath();
+    std::cout << std::endl;
+
+    TORCH_CHECK(
+        !fusion.hasDynamicTransform(), "Expected to have no dynamic transform");
+
+    fusion.printMath();
+  }
+}
+
+// Test equality of DynamicTransformInfo
+TEST_F(NVFuserTest, DynamicTransform7_CUDA) {
+  // Represents a series of reshapes
+  struct TransformList {
+    std::vector<std::vector<int64_t>> shapes;
+  };
+
+  struct ShapeInfo {
+    TransformList ref_transform;
+    std::vector<TransformList> equal_transforms;
+    std::vector<TransformList> different_transforms;
+  };
+
+  std::vector<ShapeInfo> patterns;
+
+  patterns.push_back(ShapeInfo{
+      .ref_transform = {{{3, 4}, {4, 3}}},
+      .equal_transforms =
+          {{{{3, 4}, {4, 3}}}, {{{2, 8}, {4, 4}}}, {{{3, 8}, {4, 6}}}},
+      .different_transforms = {{{{3, 4}, {2, 6}}}}});
```

Comment:
Here, this defines several patterns of reshape ops. The ref transform is used as the reference, and is compared with the equal and different transforms. The equal transforms is a list of transforms that should be able to use the same concretized fusion as the reference, whereas the different transforms should result in different concretized fusions. There are TORCH_CHECK near the end of this test that assert these hypotheses.